racket/src/cs/rumble/struct.ss

```diff
@@ -18,94 +18,113 @@
+           [st (make-struct-type-prop name (and (not (eq? guard 'can-impersonate)) guard) supers)]
+           [pred (escapes-ok
+                  (lambda (v)
+                    (let* ([v (strip-impersonator v)]
+                           [rtd (if (record-type-descriptor? v)
+                                    v
+                                    (and (record? v)
+                                         (record-rtd v)))])
+                      (and rtd
+                           (not (eq? none (struct-property-ref st rtd none)))))))]
+           [accessor-name (string->symbol (string-append
+                                           (symbol->string name)
+                                           "-ref"))]
+           [predicate-name (string->symbol
+                            (string-append
+                             (symbol->string name)
+                             "?"))]
+           [default-fail
+             (escapes-ok
+              (lambda (v)
+                (raise-argument-error accessor-name
+                                      (symbol->string predicate-name)
+                                      v)))]
+           [do-fail (lambda (fail v)
+                      (cond
+                       [(eq? fail default-fail) (default-fail v)]
+                       [(procedure? fail) (|#%app| fail)]
+                       [else fail]))])
+      (letrec ([acc
+                (case-lambda
+                 [(v fail)
+                  (cond
+                   [(and (impersonator? v)
+                         (pred v))
+                    (impersonate-struct-or-property-ref acc #f acc v #f #f)]
+                   [else
+                    (let* ([rtd (if (record-type-descriptor? v)
+                                   v
+                                    (and (record? v)
+                                         (record-rtd v)))])
+                      (if rtd
+                          (let ([pv (struct-property-ref st rtd none)])
+                            (if (eq? pv none)
+                                (do-fail fail v)
+                                pv))
+                          (do-fail fail v)))])]
+                 [(v) (acc v default-fail)])])
+        (with-global-lock*
+         (hashtable-set! property-accessors
+                         acc
+                         (cons pred can-impersonate?)))
+
+        (with-global-lock*
+         (hashtable-set! property-predicates
+                         pred
+                         st))
+        (values st
+                pred
+                acc)))]))
 
 (define (struct-type-property-accessor-procedure? v)
   (and (procedure? v)
        (let ([v (strip-impersonator v)])
          (with-global-lock* (hashtable-ref property-accessors v #f)))
        #t))
 
+(define/who struct-type-property-predicate-procedure?
+  (case-lambda
+   [(v) (struct-type-property-predicate-procedure? v #f)]
+   [(v spt) (and (procedure? v)
```

Comment:
Do you mean to allow the second argument to be `#f` as part of the public interface of `struct-type-property-predicate-procedure?`? If not, you could move the implementation to a `do-struct-type-property-predicate-procedure?` helper.